### PR TITLE
Add Split-20 - change uneven split behavior to be more torch-like

### PR DESCRIPTION
**Pull request by p-wysocki**
### Description
Introduce Split-20, which performs uneven tensor split in torch's manner: instead of reducing only the size of the last dimension of the output, it reduces the last n dimensions' sized by 1 each.

This was done using a `mode` attribute, which has 2 possible values: `torch` (default) and `legacy`. It's up to discussion if it should stay like that or maybe it should be renamed or changed to a boolean.

### Motivation and Context
Issue: https://github.com/onnx/onnx/issues/4742

### TODO
- Update reference evaluator
- Confirm the correctness of setting a 0-size dimension in ONNX (`test_split_uneven_split_3d_4`)
- Add more test cases and assert backend tests are correct

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -721,7 +723,13 @@ ONNX_OPERATOR_SET_SCHEMA(
             "If the tensor is not evenly splittable the last chunk will be smaller.",
             AttributeProto::INT,
             false)
-        .SetDoc(Split_ver18_doc)
+        .Attr(
+            "mode",
+            "Uneven split mode. "
+            "Possible values are 'torch' (default) and 'legacy'.",
+            AttributeProto::STRING,
+            std::string("torch"))
```
Comment (xadupre) [inline onnx/defs/tensor/defs.cc:731]: Should we keep the legacy as default?

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -721,7 +723,13 @@ ONNX_OPERATOR_SET_SCHEMA(
             "If the tensor is not evenly splittable the last chunk will be smaller.",
             AttributeProto::INT,
             false)
-        .SetDoc(Split_ver18_doc)
+        .Attr(
+            "mode",
+            "Uneven split mode. "
+            "Possible values are 'torch' (default) and 'legacy'.",
```
[the remark this answers is not in the document]
Comment (p-wysocki) [reply onnx/defs/tensor/defs.cc:729]: Torch, I'll change it in a minute - it was numpy before and I must have left it behind.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (p-wysocki) [reply onnx/defs/tensor/defs.cc:731]: As far as I know torch is the desired one, and I set legacy to be default in version converter when coming from opset 19 to 20. I can change it.

File: onnx/defs/tensor/defs.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (xadupre) [reply onnx/defs/tensor/defs.cc:729]: Maybe a name describring the behaviour would be better than using torch and legacy.

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -786,10 +794,22 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+                if (mode == "torch") {
+                  int reduced_dims = num_outputs * (chunk_size + 1) - split_dim_value;
+                  for (int i=0; i<num_outputs-reduced_dims; i++) {
+                    split.push_back(chunk_size+1);
+                  }
+                  while (split.size() <= num_outputs) {
+                    split.push_back(chunk_size);
+                  }
+                } else {
+                  int chunk_size = (split_dim_value / num_outputs) + 1;
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:808]: Actually, this doesn't always work. Eg., if we split 7 values into 5, then this will produce 4, 4, 4, 4, -1 which doesn't make sense. We should do 4, 3, 0, 0, 0, I think. We may need to change the description accordingly. In the other mode, we will do 2, 2, 1, 1, 1.

File: onnx/defs/tensor/defs.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:729]: Agree. May be change it to a bool (int) attribute, and call it something descriptive like "minimize_diff" (as in minimize the difference between the number of elements in the different output tensors).

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:731]: Good question. I think the current choice is ok. Especially if it is only the torch exporter that uses this, and presumably the torch converter needs the new option, not the old one.

File: onnx/defs/tensor/defs.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (p-wysocki) [reply onnx/defs/tensor/defs.cc]: Renamed to `minimize_diff`, the doc has been updated accordingly.

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -708,11 +713,15 @@ ONNX_OPERATOR_SET_SCHEMA(
...
             "num_outputs",
-            "Number of outputs to split parts of the tensor into. "
-            "If the tensor is not evenly splittable the last chunk will be smaller.",
+            "Number of outputs to split parts of the tensor into. ",
             AttributeProto::INT,
             false)
-        .SetDoc(Split_ver18_doc)
+        .Attr(
+            "minimize_diff",
+            "Uneven split mode. ",
```
Comment (xadupre) [inline onnx/defs/tensor/defs.cc:721]: extra space

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -708,11 +713,15 @@ ONNX_OPERATOR_SET_SCHEMA(
             static_cast<int64_t>(0))
         .Attr(
             "num_outputs",
-            "Number of outputs to split parts of the tensor into. "
-            "If the tensor is not evenly splittable the last chunk will be smaller.",
+            "Number of outputs to split parts of the tensor into. ",
```
Comment (xadupre) [inline onnx/defs/tensor/defs.cc:716]: extra space

File: onnx/version_converter/convert.h [hunk 6]
```diff
@@ -564,7 +564,7 @@ class DefaultVersionConverter : public BaseVersionConverter {
     registerAdapter(make_unique<CompatibleAdapter>("Size", OpSetID(18), OpSetID(19)));
 
     /******** 19 -> 20 ********/
-    registerAdapter("Split", 19, 20, SetAttribute(kmode, "legacy"));
+    registerAdapter("Split", 19, 20, SetAttribute(kminimize_diff, "legacy"));
```
Comment (gramalingam) [inline onnx/version_converter/convert.h:567]: It should be a boolean value now.

File: onnx/backend/test/case/node/split.py [hunk 7]
```diff
@@ -374,4 +382,276 @@ def export_2d_uneven_split_opset18() -> None:
...
+                [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0],
+                [8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0],
+            ]
+        ).astype(np.float32)
+
+        node = onnx.helper.make_node(
+            "Split",
+            inputs=["input"],
+            outputs=["output_1", "output_2", "output_3"],
+            axis=0,
```
Comment (gramalingam) [inline onnx/backend/test/case/node/split.py:642]: Instead of changing the `axis`, it would make sense to change the new attribute introduced `minimize_diff` or whatever it is called. At least, I am looking for some test-case to show the difference that attribute makes, whether that is this test or some other test.

**Comment by gramalingam** [comment]
Hi, is this ready for review? Would it be possible to complete this PR for the upcoming release?

**Comment by justinchuby** [comment]
@p-wysocki

File: onnx/defs/tensor/defs.cc [hunk 8]
```diff
@@ -778,7 +778,7 @@ ONNX_OPERATOR_SET_SCHEMA(
             }
           } else { // no value available for 'split'
             if (num_outputs_attr) {
-              const int num_outputs = num_outputs_attr->i();
+              const long int num_outputs = num_outputs_attr->i();
```
Comment (justinchuby) [inline onnx/defs/tensor/defs.cc:781]: prefer int64_t or size_t

Comment (p-wysocki) [reply onnx/defs/tensor/defs.cc]: Changed to int64_t, I missed that one.

**Comment by p-wysocki** [comment]
Added new tests, generated docs and test files are on the way - I'm currently having some issues with protobuf in my environment, which stops me from generating them.

**Comment by gramalingam** [comment]
It will also have to be moved to a later opset (21, I guess?)

**Comment by zhenhuaw-me** [comment]
Spec like `mode:torch` requires background of how Torch implements it.

We can make it more specific by:

1. Keep `num_outputs`, and change the semantic to match [`numpy.array_split`](https://numpy.org/doc/stable/reference/generated/numpy.array_split.html) and [`torch.tensor_split`](https://pytorch.org/docs/stable/generated/torch.tensor_split.html)

> If indices_or_sections is an integer n or a zero dimensional long tensor with value n, input is split into n sections along dimension dim. If input is divisible by n along dimension dim, each section will be of equal size, input.size(dim) / n. If input is not divisible by n, the sizes of the first int(input.size(dim) % n) sections will have size int(input.size(dim) / n) + 1, and the rest will have size int(input.size(dim) / n).

2. Introduce `split_size`, which is the same as [`torch.split`](https://pytorch.org/docs/stable/generated/torch.split.html):

> If split_size_or_sections is an integer type, then [tensor](https://pytorch.org/docs/stable/generated/torch.tensor.html#torch.tensor) will be split into equally sized chunks (if possible). Last chunk will be smaller if the tensor size along the given dimension dim is not divisible by split_size.
